### ISSUE TITLE
"No decorator applied" when testing with pcf dev

### ISSUE THREAD

**Issue by ngscchung** [cf-platform-eng/meta-buildpack#5]
I'm running pcf dev locally and created meta-buildpack and test_decorator, then when I tried to push the spring-music app, I can see the "[meta-buildpack] compiling..." logs which triggers the java-buildpack, but I can't see the "[test-decorator] compiling...", and at the end of the cf push log I saw the "... (no decorators apply)" statement... tried to set the test_decorator priority to "2" but still no luck... Below shared some logs. Any helps are appreciated.

$ cf buildpacks
Getting buildpacks...

buildpack                    position   enabled   locked   filename
meta_buildpack               1          true      false    meta-buildpack-master.zip
java_buildpack               2          true      false    java-buildpack-offline-v3.8.1.zip
ruby_buildpack               3          true      false    ruby_buildpack-cached-v1.6.19.zip
nodejs_buildpack             4          true      false    nodejs_buildpack-cached-v1.5.15.zip
go_buildpack                 4          true      false    go_buildpack-cached-v1.7.10.zip
python_buildpack             5          true      false    python_buildpack-cached-v1.5.7.zip
php_buildpack                6          true      false    php_buildpack-cached-v4.3.14.zip
staticfile_buildpack         7          true      false    staticfile_buildpack-cached-v1.3.9.zip
binary_buildpack             7          true      false    binary_buildpack-cached-v1.0.2.zip
dotnet_core_buildpack_beta   8          true      false    dotnet-core-buildpack.zip
test_decorator               9          true      false    test_decorator.zip

cf push logs:
...
Downloading build artifacts cache...
Downloaded build artifacts cache (109B)
Staging...
[meta-buildpack] Compiling with java-buildpack=v3.8.1-offline-https://github.com/cloudfoundry/java-buildpack.git#29c79f2 java-main open-jdk-like-jre=1.8.0_91-unlimited-crypto open-jdk-like-memory-calculator=2.0.2_RELEASE spring-auto-reconfiguration=1.10.0_RELEASE
...
Showing health and status for app spring-music in org pcfdev-org / space pcfdev-space as admin...
OK

requested state: started
instances: 1/1
usage: 1G x 1 instances
urls: spring-music-atrophied-latimeria.local.pcfdev.io
last uploaded: Tue May 2 11:32:05 UTC 2017
stack: cflinuxfs2
buildpack: java-buildpack=v3.8.1-offline-https://github.com/cloudfoundry/java-buildpack.git#29c79f2 java-main open-jdk-like-jre=1.8.0_91-unlimited-crypto open-jdk-like-memory-calculator=2.0.2_RELEASE spring-auto-reconfiguration=1.10.0_RELEASE (no decorators apply)

**Comment by guidowb** [comment]
Couple of things to check:
- Does your decorator have a bin/decorate script?
- Is that script executable?
- Is it actually being executed? (write something to stderr and it should show up in the log)
- Does it exit zero?

**Comment by ngscchung** [comment]
I found what's my mistakes ... my test-decorator is actually the one included in [meta-buildpack/test](https://github.com/cf-platform-eng/meta-buildpack/tree/master/test/test_decorator) folder, and the "decorate" script included condition checking for that test_app folder, plus I misunderstood that the flow was detect > compile > decorate but actually it should be detect > decorate > compile instead... I got the echo statements after removing those condition checking. Thanks for your help

**Comment by ngscchung** [comment]
Sorry but another question... I tested okay with Spring-music and another Python app... but when I try with node.js app ... the meta-buildpack and decorator seems not triggered as I can't found related logs... any clues?

**Comment by guidowb** [comment]
The only reason meta-buildpack won't get triggered is if you are explicitly specifying a buildpack, either on the `cf push` command, or in a manifest.yml. Is there a manifest file in your node app directory?

**Comment by ngscchung** [comment]
Yes you are right... found that the manifest.yml in my node project specified the nodejs_buildpack. The decorator works if I comment out that in the yml... Is it possible to make it work even if I specified buildpack in yml? or any possible workaround?

**Comment by guidowb** [comment]
Why do you need to specify the buildpack? Are you using a custom one?

**Comment by ngscchung** [comment]
One use case is custom buildpack, for e.g. we have a customized nodejs buildpack with oracle client (as specified in this [link](https://github.com/cloudfoundry/nodejs-buildpack/issues/65)). Another case is that our PCF environment would have multiple versions of nodejs or python buildpacks for backward compatibility and our app team can specify which version to use. One more case is that our deployment pipeline may read the manifest and trigger specific tasks like nodejs/python specific code scan/review.

**Comment by guidowb** [comment]
Ok, those all make sense. Unfortunately, as soon as you specify a buildpack, the staging process will invoke that buildpack directly, and meta-buildpack never gets control. So there's no way for it to take part in the staging process in that case.
I suppose you could build your own meta-buildpack variant that you can somehow tell which other buildpack to run. But that somehow would have to be outside of the manifest, so it would change the way your app team pushes their apps.
Maybe take a look at https://github.com/cloudfoundry-incubator/multi-buildpack and see if that meets your needs?

**Comment by ngscchung** [comment]
I did aware of multipack, but the documentation said that it "will not work with system buildpacks" which are being used in most of our app, so we haven't tested it. 

Sorry but I can't quite get your idea about the "meta-buildpack variant" ... and why that would "have to be outside of the manifest"... would you please elaborate more? Thanks in advance

**Comment by guidowb** [comment]
If you specify buildpacks in the manifest, meta-buildpack will not be invoked. So if you want to devise a way to tell meta-buildpack what other buildpacks to run, it will have to be outside of the manifest. It can't be in the standard place.

And you would then have to modify meta-buildpack to look at that information and run the specified buildpacks, instead of doing the auto-detect it does today. I think it might not be easy, because meta-buildpack gets very little information about the other buildpacks (it doesn't, for instance, know their names). But I haven't looked at it in a while.

If you decide to go down that route, I'm happy to work with you on the design and accept a PR to meta-buildpack.
